[flang] NAMLIST group should not accept parameter variable

Here is an example of compilers emitting an error, while flang does not:
```
$ cat c3.f90
program p
  implicit none
  integer, parameter :: k = 3
  namelist /g/ k
end program

$ gfortran c3.f90
c3.f90:4:16:

    4 |   namelist /g/ k
      |                1
Error: PARAMETER attribute conflicts with NAMELIST attribute in ‘k’ at (1)

$ crayftn c3.f90

  namelist /g/ k
               ^ 
ftn-552 ftn: ERROR P, File = c3.f90, Line = 4, Column = 16 
  "K" has the PARAMETER attribute, therefore it must not be declared as a namelist-group-object.

Cray Fortran : 5 source lines
Cray Fortran : 1 errors, 0 warnings, 0 other messages, 0 ansi
Cray Fortran : "explain ftn-message number" gives more information about each message.

$ flang c3.f90
$ 
```

A NAMELIST group can be used for both input and output, and there is no problem with having undefinable items in a NAMELIST group that is never used for input.  If you do try to use a named constant or INTENT(IN) dummy argument in a NAMELIST group in a READ statement, you will observe errors.

```
λ cat t.f90
subroutine s(x)
  integer, intent(in) :: x
  integer, parameter :: y = 123
  namelist /nml/ x, y
  read(*,nml=nml)
end
λ flang-new -c t.f90
error: Semantic errors in t.f90
./t.f90:5:14: error: NAMELIST input group must not contain undefinable item 'x'
    read(*,nml=nml)
               ^^^
./t.f90:5:14: because: 'x' is an INTENT(IN) dummy argument
    read(*,nml=nml)
               ^^^
./t.f90:2:26: Declaration of 'x'
    integer, intent(in) :: x
                           ^
./t.f90:5:14: error: NAMELIST input group must not contain undefinable item 'y'
    read(*,nml=nml)
               ^^^
./t.f90:5:14: because: 'y' is not a variable
    read(*,nml=nml)
               ^^^
./t.f90:3:25: Declaration of 'y'
    integer, parameter :: y = 123
                          ^
```

--- a/flang/docs/Extensions.md
+++ b/flang/docs/Extensions.md
@@ -484,6 +484,18 @@ end
 * A pointer component that has no default initialization or explicit value
   in a structure constructor is defaulted to `NULL()`.
 * An assumed-rank entity is an acceptable `NAMELIST` group item.
+* A named constant (`PARAMETER`) may appear as a `namelist-group-object` in a
+  `NAMELIST` statement.  The Fortran standard requires namelist group objects
+  to be variables, but this usage is accepted by Flang as an extension.
+  When `-pedantic` is enabled, Flang emits a warning for this case.
+  For example:
+```
+program p
+  implicit none
+  integer, parameter :: k = 3
+  namelist /g/ k
+end program
+```
 
 ### Extensions supported when enabled by options
 

--- a/flang/include/flang/Support/Fortran-features.h
+++ b/flang/include/flang/Support/Fortran-features.h
@@ -82,7 +82,7 @@ ENUM_CLASS(UsageWarning, Portability, PointerToUndefinable,
     HostAssociatedIntentOutInSpecExpr, NonVolatilePointerToVolatile,
     RealConstantWidening, VolatileOrAsynchronousTemporary, UnusedVariable,
     UsedUndefinedVariable, BadValueInDeadCode, AssumedTypeSizeDummy,
-    MisplacedIgnoreTKR)
+    MisplacedIgnoreTKR, NamelistParameter)
 
 using LanguageFeatures = EnumSet<LanguageFeature, LanguageFeature_enumSize>;
 using UsageWarnings = EnumSet<UsageWarning, UsageWarning_enumSize>;

--- a/flang/lib/Semantics/check-namelist.cpp
+++ b/flang/lib/Semantics/check-namelist.cpp
@@ -7,6 +7,7 @@
 //===----------------------------------------------------------------------===//
 
 #include "check-namelist.h"
+#include "flang/Semantics/tools.h"
 
 namespace Fortran::semantics {
 
@@ -26,6 +27,13 @@ void NamelistChecker::Leave(const parser::NamelistStmt &nmlStmt) {
             context_.Say(nmlObjName.source,
                 "A PRIVATE namelist group object '%s' must not be in a "
                 "PUBLIC namelist"_err_en_US,
+                nmlObjSymbol->name());
+          }
+          // `namelist-group-object` may only contain variables.
+          if (IsNamedConstant(*nmlObjSymbol)) {
+            context_.Warn(common::UsageWarning::NamelistParameter,
+                nmlObjName.source,
+                "A namelist group object '%s' should not be a PARAMETER"_port_en_US,
                 nmlObjSymbol->name());
           }
         }

--- a/flang/test/Semantics/namelist02.f90
+++ b/flang/test/Semantics/namelist02.f90
@@ -1,0 +1,29 @@
+! RUN: %python %S/test_errors.py %s %flang_fc1 -pedantic -Werror
+
+module m
+  implicit none
+  integer, parameter :: mc = 42
+end module
+
+! Local named constant
+program p
+  use m
+  implicit none
+  integer, parameter :: k = 3
+  !PORTABILITY: A namelist group object 'k' should not be a PARAMETER [-Wnamelist-parameter]
+  namelist /g/ k
+  ! USE-associated named constant
+  !PORTABILITY: A namelist group object 'mc' should not be a PARAMETER [-Wnamelist-parameter]
+  namelist /g2/ mc
+end program
+
+! Host-associated named constant
+subroutine host
+  implicit none
+  integer, parameter :: hc = 10
+  contains
+    subroutine inner
+      !PORTABILITY: A namelist group object 'hc' should not be a PARAMETER [-Wnamelist-parameter]
+      namelist /g3/ hc
+    end subroutine
+end subroutine

